File: wpn-3-claude/src/services/DataLoader.ts
import { Attraction, WaitingTime, Genre } from '../types/Models';

/**
 * JSONデータをロードしてAttractionオブジェクトに変換
 */
export async function loadAttractions(): Promise<Attraction[]> {
  try {
    const attractionsData = require('../../assets/data/attractions.json');

    return attractionsData.map((data: any) => ({
      id: data.id,
      name: data.name,
      translatedName: data.translated_name,
      officialId: data.official_id,
      entranceLat: data.entrance_lat,
      entranceLng: data.entrance_lng,
      exitLat: data.exit_lat,
      exitLng: data.exit_lng,
      areaName: data.area_name,
      genre: data.genre as Genre,
      icon: data.icon,
      durationMinutes: data.duration_minutes,
      isSeated: data.is_seated,
      isActive: data.is_active,
      isInvalid: data.is_invalid,
    }));
  } catch (error) {
    console.error('Failed to load attractions:', error);
    return [];
  }
}

/**
 * JSONデータをロードしてWaitingTimeオブジェクトに変換
 */
export async function loadWaitingTimes(): Promise<WaitingTime[]> {
  try {
    const waitingTimesData = require('../../assets/data/waiting_times.json');

    return waitingTimesData.map((data: any) => ({
      attrId: data.attr_id,
      waitingMinutes: data.waiting_minutes,
      updatedAt: data.updated_at,
      timeSeries: data.time_series.map((ts: any) => ({
        timestamp: ts.timestamp,
        waitingMinutes: ts.waiting_minutes,
      })),
    }));
  } catch (error) {
    console.error('Failed to load waiting times:', error);
    return [];
  }
}

/**
 * 待ち時間データをMapに変換
 */
export function createWaitingTimesMap(
  waitingTimes: WaitingTime[]
): Map<string, WaitingTime> {
  const map = new Map<string, WaitingTime>();
  waitingTimes.forEach((wt) => {
    map.set(wt.attrId, wt);
  });
  return map;
}

File: wpn-1/wpn-1/DataLoader.swift
import Foundation

final class DataLoader {
    static let shared = DataLoader()

    private(set) var attractions: [Attraction] = []
    private(set) var waitingTimeRecords: [WaitingTimeRecord] = []

    private init() {}

    func loadInitialData() {
        loadAttractions()
        mergeGreetings()
        loadWaitingTimes()
    }

    private func loadAttractions() {
        guard let url = Bundle.main.url(forResource: "attractions", withExtension: "json") else {
            print("[DataLoader] attractions.json not found in bundle")
            return
        }
        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            attractions = try decoder.decode([Attraction].self, from: data)
        } catch {
            print("[DataLoader] Failed to decode attractions.json: \(error)")
        }
    }

    /// greetings.json からグリーティング施設を読み込み、アトラクション一覧にマージ
    private func mergeGreetings() {
        guard let url = Bundle.main.url(forResource: "greetings", withExtension: "json") else {
            print("[DataLoader] greetings.json not found in bundle")
            return
        }
        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let greetings = try decoder.decode([GreetingAttraction].self, from: data)

            var maxId = attractions.map { $0.id }.max() ?? 0
            let existingOfficialIds = Set(attractions.compactMap { $0.officialId })

            for g in greetings {
                // すでに同じ officialId のアトラクションがあればスキップ
                if existingOfficialIds.contains(g.officialId) { continue }

                maxId += 1
                let attraction = Attraction(
                    id: maxId,
                    name: g.name,
                    officialId: g.officialId,
                    entranceLat: g.lat,
                    entranceLng: g.lng,
                    exitLat: g.lat,
                    exitLng: g.lng,
                    areaName: g.areaName,
                    isActive: true,
                    isInvalid: false
                )
                attractions.append(attraction)
            }

            print("[DataLoader] Merged \(greetings.count) greeting attractions")
        } catch {
            print("[DataLoader] Failed to decode greetings.json: \(error)")
        }
    }

    private func loadWaitingTimes() {
        guard let url = Bundle.main.url(forResource: "waiting_times", withExtension: "json") else {
            print("[DataLoader] waiting_times.json not found in bundle")
            return
        }
        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            decoder.dateDecodingStrategy = .formatted(DateFormatter.waitingTimeISO8601)
            waitingTimeRecords = try decoder.decode([WaitingTimeRecord].self, from: data)
        } catch {
            print("[DataLoader] Failed to decode waiting_times.json: \(error)")
        }
    }

    /// 到着予定時刻に対して最も近い待ち時間データを返す（分と元データの時刻）
    func waitingTimeDetail(for attraction: Attraction, arrivalMinutes: Int) -> (minutes: Int, sourceTime: Date?) {
        guard
            let officialId = attraction.officialId,
            let officialInt = Int(officialId),
            let record = waitingTimeRecords.first(where: { $0.attrId == officialInt })
        else {
            return (attraction.waitingMinutes, nil)
        }

        // 日内分数で最も近い timeSeries を探す
        if !record.timeSeries.isEmpty {
            let calendar = Calendar(identifier: .gregorian)
            let targetDate = calendar.startOfDay(for: record.updatedAt).addingTimeInterval(TimeInterval(arrivalMinutes * 60))
            let nearest = record.timeSeries.min(by: { a, b in
                abs(a.timestamp.timeIntervalSince(targetDate)) < abs(b.timestamp.timeIntervalSince(targetDate))
            })
            if let nearest = nearest {
                return (nearest.waitingMinutes, nearest.timestamp)
            }
        }

        return (record.waitingMinutes, record.updatedAt)
    }

    /// 既存ロジック互換: 待ち時間（分）のみを返す
    func waitingTime(for attraction: Attraction, arrivalMinutes: Int) -> Int {
        waitingTimeDetail(for: attraction, arrivalMinutes: arrivalMinutes).minutes
    }
}

